### PR TITLE
prometheus codec for label names and values; apply to cache key generator

### PR DESCRIPTION
**Pull request by francoposa**
Additions to the Prometheus Codec, per https://github.com/grafana/mimir-squad/issues/1938

A fair amount of this logic for processing Label Names and Label Values requests already existed at in various areas of the code, particularly in `pkg/frontend/querymiddleware/labels_query_cache.go`, but was not encapsulated into interfaces or types to represent the Label Names and Label Values requests.

This PR kind of looks like we're just moving stuff around, but the goal is to eventually have everything work with the codec types instead of just `http.Request` so that we don't have to re-parse the interesting data out of the request several times.

#### What this PR does

* Creates `LabelsQueryRequest` interface type with concrete implementations in protos `PrometheusLabelNamesQueryRequest` and `PrometheusLabelValuesQueryRequest`; mirroring the approach for the interface `Request` and its concrete types for Range and Instant queries
* Implements Encode & Decode to/from the bare `http.Request` and adds these interfaces to the prometheus Codec interface we use
* updates the `keyingFunc` type used for cache keys to just take a bare `http.Request` instead of the chopped up (ctx, path, url.Values) pieces of it. Since the Prometheus codec takes the approach of the interface always encoding to/from the `http.Request`, it is easier to let the codec do the work instead of the caller always chopping up the request before passing to a `keyingFunc`.

I also intend to have a follow-on PR for pure renaming of the `Request` interface type to `MetricsQueryRequest`.
It is the original type used for the Prometheus codec but was only ever designed and intended to represent Range and Instant queries. Now that we have another type for labels, we need to disambiguate.

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir-squad/issues/1941

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/frontend/querymiddleware/model_extra.go [hunk 1]
```diff
@@ -161,6 +162,54 @@ func (r *PrometheusInstantQueryRequest) AddSpanTags(sp opentracing.Span) {
...
+	if r.GetEnd() == 0 {
+		return v1.MaxTime.UnixMilli()
+	}
+	return r.GetEnd()
+}
+
+// AddSpanTags writes query information about the current `PrometheusLabelNamesQueryRequest`
+// to a span's tag ("attributes" in OpenTelemetry parlance).
+func (r *PrometheusLabelNamesQueryRequest) AddSpanTags(sp opentracing.Span) {
+	//sp.SetTag("query", r.GetQuery())
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/model_extra.go:200]: ?

File: pkg/frontend/querymiddleware/codec.go [hunk 2]
```diff
@@ -409,6 +496,53 @@ func (c prometheusCodec) EncodeRequest(ctx context.Context, r Request) (*http.Re
 	return req.WithContext(ctx), nil
 }
 
+func (c prometheusCodec) EncodeLabelsQueryRequest(ctx context.Context, req LabelsQueryRequest) (*http.Request, error) {
+	var u *url.URL
+	switch req := req.(type) {
+	case *PrometheusLabelNamesQueryRequest:
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go:502]: I don't understand why this can only encode one of the implementations of `LabelsQueryRequest`

**Review by 56quarters — COMMENTED** [review]
Looks good overall, thanks!

File: pkg/frontend/querymiddleware/model_extra.go (hunk 1 above)
Comment (francoposa) [reply pkg/frontend/querymiddleware/model_extra.go:200]: ah yeah, my mistake here - intended do have a question on whether I should just use the "query" span tag for consistency and just put the matchers in there, or use a new span tag.

File: pkg/frontend/querymiddleware/codec.go (hunk 2 above)
Comment (francoposa) [reply pkg/frontend/querymiddleware/codec.go:502]: whoop yeah missed that as I originally only intended to PR for the Label Names but ended up adding Label Values as well... coming up

File: pkg/frontend/querymiddleware/codec.go [hunk 3]
```diff
@@ -256,6 +285,39 @@ func (c prometheusCodec) decodeInstantQueryRequest(r *http.Request) (Request, er
 	return &result, nil
 }
 
+func (prometheusCodec) DecodeLabelsQueryRequest(_ context.Context, r *http.Request) (LabelsQueryRequest, error) {
+	if !IsLabelsQuery(r.URL.Path) {
+		return nil, fmt.Errorf("unknown labels query API endpoint %s", r.URL.Path)
+	}
+	start, end, err := DecodeLabelsQueryTimeParams(r, false)
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/frontend/querymiddleware/codec.go]: converted all `Decode[Something]TimeParams` to this as they all had the same issue.

File: pkg/frontend/querymiddleware/codec.go [hunk 4]
```diff
@@ -115,6 +119,31 @@ type Request interface {
...
+	GetEnd() int64
+	// GetEndOrDefault returns the end timestamp of the request in milliseconds,
+	// or the Prometheus v1 API MaxTime if no end timestamp was provided on the original request.
+	GetEndOrDefault() int64
+	// GetLabelMatcherSets returns the label matchers a.k.a series selectors for Prometheus label query requests,
+	// as retained in their original string format. This enables the request to be symmetrically decoded and encoded
+	// to and from the http request format without needing to undo the Prometheus parser converting between formats
+	// like `up{job="prometheus"}` and `{__name__="up, job="prometheus"}`, or other idiosyncrasies.
+	GetLabelMatcherSets() []string
+	proto.Message
```
[the remark this answers is not in the document]
Comment (francoposa) [reply pkg/frontend/querymiddleware/codec.go]: done

File: pkg/frontend/querymiddleware/model_extra.go (hunk 1 above)
Comment (francoposa) [reply pkg/frontend/querymiddleware/model_extra.go]: ended up using "matchers" as key as it is used on KV logging elsewhere in Mimir/Prometheus

File: pkg/frontend/querymiddleware/codec.go (hunk 2 above)
Comment (francoposa) [reply pkg/frontend/querymiddleware/codec.go:518]: added, with test cases

**Review by 56quarters — APPROVED** [review]
LGTM, thank you!

**Pull request merged**